ADR/0031-sprayproxy.md

```diff
@@ -0,0 +1,29 @@
+# 0031 Sprayproxy
+
+Created: 2023-10-24
+
+## Status
+
+Proposed
+
+## Context
+
+AppStudio has multiple member (backend) clusters. Each member cluster is running a Pipelines-As-Code (PaC) service, accepting webhook requests. A GitHub App can only specify a single destination for webhook requests. We need to forward those requests to multiple clusters.
+
+## Decision
+
+Deploy a service (`Sprayproxy`) on the AppStudio host (frontend) clusters. The service route is configured in the GitHub App as a `Webhook URL`, so all webhook requests are directed to it. The service has a list of backends configured. The service does not distinguish between the type of requests the way PaC does (pull-request/push/comment etc), it treats them all equally. For each incoming request, a new outgoing request is constructed with the original payload and destination of each of the member clusters.
+
+The service performs the following checks on incoming requests:
+
+- Validating the request is from GitHub using a shared secret. The process is officially documented [here](https://docs.github.com/en/webhooks/using-webhooks/validating-webhook-deliveries).
+- [Payloads are capped at 25 MB](https://docs.github.com/en/webhooks/webhook-events-and-payloads#payload-cap), but that's configurable.
+
+The QE team requires the ability to dynamically add/remove backends. That functionality is disabled by default as it represents a security risk.
```

Comment:
I propose to expand this paragraph a little.
1. Explain where the security risks come from (secret to access the repo is part of payload, bad actor could access private content if they get the forwarded payload).
2. Explain a little bit on how the registration/unregistration mechanism works.

Reply:
Added.